mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -604,6 +609,42 @@ def log_metric(self, run_id, metric):
             if just_created:
                 self._update_latest_metric_if_necessary(logged_metric, session)
 
+    def _log_metrics(self, run_id, metrics, session=None):
+        if not metrics:
+            return
+
+        # Duplicate metric values are eliminated here to maintain
+        # the same behavior in log_metric
+        metric_instances = []
+        seen = set()
+        for metric in metrics:
+            metric, value, is_nan = self._get_metric_value_details(metric)
+            if metric not in seen:
+                metric_instances.append(SqlMetric(
+                    run_uuid=run_id,
+                    key=metric.key,
+                    value=value,
+                    timestamp=metric.timestamp,
+                    step=metric.step,
+                    is_nan=is_nan
+                ))
+            seen.add(metric)
+
+        # use nullcontext approach if a session is given, otherwise obtain
+        # a new session via ManagedSessionMaker
+        context = (
+            self.ManagedSessionMaker() if session is None
+            else contextlib.nullcontext(session)
+        )
+
+        with context as session:
+            self._save_to_db(session=session, objs=metric_instances)
```

Comment:
It just occurred to me that we may run into trouble with the idempotence of LogBatch due to this use of `session.add_all()` (`_save_to_db()`) to insert a collection of new `Metric` records into the `metrics` table. If two LogBatch requests with the same metric content are issued sequentially, the second one will fail with a primary key constraint violation.

Can we implement logic similar to what we're using in `_log_params` by catching `IntegrityError` and checking the run's metric history? If the metric history contains all of the metrics being inserted, we can simply return, since we know that the `metrics` table contains the desired values and that they were reflected correctly by `latest_metrics` when they were originally inserted.

Reply:
Yeah, it is a good idea. If the metric history does not contain all of the metrics being inserted, I should insert the ones whom not present, and update latest_metrics table, is it correct?

Reply:
Yep! Exactly :).

Reply:
I think this one is done